File: ASSIGNMENT10/python/Q.10.py
class Graph:
    def __init__(self):
        self.adjacency_list = {}

    def add_edge(self, u, v):
        if u not in self.adjacency_list:
            self.adjacency_list[u] = []
        if v not in self.adjacency_list:
            self.adjacency_list[v] = []
        self.adjacency_list[u].append(v)
        self.adjacency_list[v].append(u)  # Uncomment this line for undirected graph

    def dfs(self, start, visited=None):
        if visited is None:
            visited = set()  # Set to keep track of visited nodes
        visited.add(start)  # Mark the current node as visited
        dfs_order = [start]  # List to store the DFS traversal order

        # Recur for all the vertices adjacent to this vertex
        for neighbor in self.adjacency_list.get(start, []):
            if neighbor not in visited:
                dfs_order.extend(self.dfs(neighbor, visited))

        return dfs_order

# Example Usage
if __name__ == "__main__":
    g = Graph()
    
    # Adding edges
    g.add_edge(0, 1)
    g.add_edge(0, 2)
    g.add_edge(1, 3)
    g.add_edge(1, 4)
    g.add_edge(2, 5)
    g.add_edge(2, 6)
    
    # Perform DFS starting from vertex 0
    dfs_result = g.dfs(0)

    # Print the result of DFS
    print("DFS Traversal starting from vertex 0:", dfs_result)

File: ASSIGNMENT10/c/Q.10.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Graph {
    int vertices;
    struct Node** adjLists;
};

struct Node* createNode(int v) {
    struct Node* newNode = malloc(sizeof(struct Node));
    newNode->data = v;
    newNode->next = NULL;
    return newNode;
}

struct Graph* createGraph(int vertices) {
    struct Graph* graph = malloc(sizeof(struct Graph));
    graph->vertices = vertices;

    graph->adjLists = malloc(vertices * sizeof(struct Node*));

    for (int i = 0; i < vertices; i++) {
        graph->adjLists[i] = NULL;
    }

    return graph;
}

void addEdge(struct Graph* graph, int src, int dest) {
    struct Node* newNode = createNode(dest);
    newNode->next = graph->adjLists[src];
    graph->adjLists[src] = newNode;
}

void dfs(struct Graph* graph, int vertex, int* visited) {
    visited[vertex] = 1; // Mark as visited
    printf("Visited %d\n", vertex);

    struct Node* temp = graph->adjLists[vertex];
    while (temp) {
        int adjVertex = temp->data;
        if (!visited[adjVertex]) {
            dfs(graph, adjVertex, visited);
        }
        temp = temp->next;
    }
}

int main() {
    struct Graph* graph = createGraph(5);
    addEdge(graph, 0, 1);
    addEdge(graph, 0, 4);
    addEdge(graph, 1, 2);
    addEdge(graph, 1, 3);
    addEdge(graph, 1, 4);
    addEdge(graph, 2, 3);
    addEdge(graph, 3, 4);

    int visited[5] = {0};

    printf("Depth First Search starting from vertex 0:\n");
    dfs(graph, 0, visited);
    return 0;
}
